neqo-http3/src/connection.rs

```diff
@@ -901,2519 +703,178 @@ impl Http3Connection {
     }
+}
 
-    // SERVER SIDE ONLY FUNCTIONS
-    fn handle_new_client_request(&mut self, stream_id: u64) {
-        self.transactions_server
-            .insert(stream_id, TransactionServer::new(stream_id));
+impl ::std::fmt::Display for Http3ClientHandler {
+    fn fmt(&self, f: &mut ::std::fmt::Formatter) -> ::std::fmt::Result {
+        write!(f, "Http3 connection client")
     }
 }
 
-#[derive(Debug, PartialOrd, Ord, PartialEq, Eq, Clone)]
-pub enum Http3Event {
-    /// Space available in the buffer for an application write to succeed.
-    HeaderReady { stream_id: u64 },
-    /// A stream can accept new data.
-    DataWritable { stream_id: u64 },
-    /// New bytes available for reading.
-    DataReadable { stream_id: u64 },
-    /// Peer reset the stream.
-    Reset { stream_id: u64, error: AppError },
-    /// Peer has send STOP_SENDING with error code EarlyResponse, other error will post a reset event.
-    StopSending { stream_id: u64, error: AppError },
-    /// A new push stream
-    NewPushStream { stream_id: u64 },
-    /// New stream can be created
-    RequestsCreatable,
-    /// Cert authentication needed
-    AuthenticationNeeded,
-    /// Client has received a GOAWAY frame
-    GoawayReceived,
-    /// Connection state change.
-    StateChange(Http3State),
-}
+#[derive(Default)]
+pub struct Http3ServerHandler {}
 
-#[derive(Debug, Default, Clone)]
-pub struct Http3Events {
-    events: Rc<RefCell<BTreeSet<Http3Event>>>,
-}
-
-impl Http3Events {
-    pub fn header_ready(&self, stream_id: u64) {
-        self.insert(Http3Event::HeaderReady { stream_id });
-    }
-
-    pub fn data_writable(&self, stream_id: u64) {
-        self.insert(Http3Event::DataWritable { stream_id });
-    }
-
-    pub fn data_readable(&self, stream_id: u64) {
-        self.insert(Http3Event::DataReadable { stream_id });
-    }
-
-    pub fn reset(&self, stream_id: u64, error: AppError) {
-        self.insert(Http3Event::Reset { stream_id, error });
+impl Http3Handler<Http3ServerEvents, TransactionServer> for Http3ServerHandler {
+    fn new() -> Self {
+        Http3ServerHandler::default()
     }
 
-    pub fn stop_sending(&self, stream_id: u64, error: AppError) {
-        self.insert(Http3Event::StopSending { stream_id, error });
-    }
-
-    pub fn new_push_stream(&self, stream_id: u64) {
-        self.insert(Http3Event::NewPushStream { stream_id });
-    }
-
-    pub fn new_requests_creatable(&self) {
-        self.insert(Http3Event::RequestsCreatable);
-    }
-
-    pub fn authentication_needed(&self) {
-        self.insert(Http3Event::AuthenticationNeeded);
-    }
-
-    pub fn goaway_received(&self) {
-        self.insert(Http3Event::GoawayReceived);
-    }
-
-    pub fn connection_state_change(&self, state: Http3State) {
-        self.insert(Http3Event::StateChange(state));
+    fn handle_new_bidi_stream(
+        &mut self,
+        transactions: &mut HashMap<u64, TransactionServer>,
+        _conn: &mut Connection,
```

Comment:
this is because server and client behaves differently. The server adds a new transaction and server closes the stream, one need 'transactions' the other 'conn'.